File: Spark_Datastream/task2.scala
import scala.io.Source
import scala.util.Random
import scala.collection.mutable
import java.io.PrintWriter
import java.io.File
import Blackbox._  

object task2 {
  val BIT_ARRAY_SIZE: Int = 69997 
  val NUM_HASH_FUNCTIONS: Int = 10

  //   f(x) = ((a * x + b) % p) % m
  def generateHash(m: Int = BIT_ARRAY_SIZE, numHush: Int = NUM_HASH_FUNCTIONS): List[BigInt => Int] = {


    val primeList = List(1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069)
    (1 to numHush).map { _ =>
      val p = primeList(Random.nextInt(primeList.length))
      val a = Random.nextInt(p - 1) + 1 // random integer in [1, p-1]
      val b = Random.nextInt(p - 1) + 1
      (x: BigInt) => (((a * x + b) % p) % m).toInt
    }.toList
  }


  val globalHashFunctions: List[BigInt => Int] = generateHash()


  def myhashs(s: String): Seq[Int] = {
    val inputNumber = BigInt(s.getBytes("UTF-8"))
    globalHashFunctions.map(hf => hf(inputNumber))
  }


  def countTrailingZeros(x: Int): Int = {
    var count = 0
    var y = x
    if (y == 0) return 0 // safeguard; ideally, x should not be 0.
    while (y % 2 == 0) {
      count += 1
      y = y / 2
    }
    count
  }

  def main(args: Array[String]): Unit = {
    if (args.length != 4) {
      println("Usage: Task1 <input_file> <stream_size> <num_of_asks> <output_file>")
      sys.exit(1)
    }

    val inputFile = args(0)
    val streamSize = args(1).toInt
    val numOfAsks = args(2).toInt
    val outputFile = args(3)

    val bx = Blackbox()
    
    // Open output CSV file.
    val pw = new PrintWriter(new File(outputFile))
    pw.println("Time,Ground Truth,Estimation")
    
    var totalEstimation: Double = 0.0
    var totalGroundTruth: Double = 0.0

    for (i <- 0 until numOfAsks) {

      val stream: Array[String] = bx.ask(inputFile, streamSize)
      val groundTruth = stream.distinct.length
      totalGroundTruth += groundTruth

      val R: Array[Int] = Array.fill(NUM_HASH_FUNCTIONS)(0)
      
  
      for (user <- stream) {
        val hvals: Seq[Int] = myhashs(user)
        hvals.zipWithIndex.foreach { case (h, j) =>
          val tz = countTrailingZeros(h)
          if (tz > R(j)) R(j) = tz
        }
      }
      
      
      val estimates: Seq[Double] = R.map(r => math.pow(2, r))
      


      val groupSize = 10
      val groupAvgs: Seq[Double] = estimates.grouped(groupSize).map { group =>
        group.sum / group.size.toDouble
      }.toSeq.sorted
      

      val n = groupAvgs.length
      val fmEstimate: Double = 
        if (n % 2 == 1) groupAvgs(n / 2)
        else (groupAvgs(n / 2 - 1) + groupAvgs(n / 2)) / 2.0
      
      // Convert fmEstimate to an integer by rounding.
      val fmEstimateInt: Int = math.round(fmEstimate).toInt
      
      totalEstimation += fmEstimateInt
      
      pw.println(s"$i,$groundTruth,$fmEstimateInt")
    }

    pw.close()

    if (totalGroundTruth > 0) {
      println(s"\nOverall ratio (sum of estimations / sum of ground truths): ${totalEstimation / totalGroundTruth}")
    } else {
      println("\nNo ground truth values were accumulated.")
    }
  }
}

File: Spark_Datastream/task2.py
from blackbox import BlackBox
import sys
import random
import binascii

BIT_ARRAY_SIZE = 69997
NUM_HASH_FUNCTIONS = 10  # Using 10 hash functions

# ((ax + b) % p) % m
def generate_hash(m = BIT_ARRAY_SIZE, num_hush = NUM_HASH_FUNCTIONS):

    prime_list = [997, 991, 983, 977, 971, 967, 953, 947, 941, 937]

    hash_functions = []
    for _ in range(num_hush):
        p = random.choice(prime_list)
        a = random.randint(1, p-1)
        b = random.randint(1, p-1)
        def hash_func(x, a=a, b=b, p=p):
            return ((a * x + b) % p) % m
        hash_functions.append(hash_func)
    return hash_functions

global_hash_functions = generate_hash()

# get the result of hash 
def myhashs(s):
    hash_results = []
    input_number = int(binascii.hexlify(s.encode('utf8')), 16)
    for hash_func in global_hash_functions:
        hash_results.append(hash_func(input_number))
    return hash_results

def main(input_file, stream_size, num_of_asks, output_file):
    bx = BlackBox()

    def count_trailing_zeros(x):
        if x == 0:
            return 0  
        count = 0
        while x % 2 == 0:
            count += 1
            x //= 2
        return count

    total_estimation = 0.0
    total_ground_truth = 0.0

    with open(output_file, 'w') as f_out:

        f_out.write("Time,Ground Truth,Estimation\n")
        
        for i in range(num_of_asks):
            stream = bx.ask(input_file, stream_size)
            ground_truth = len(set(stream))
            total_ground_truth += ground_truth
            
            R = [0] * NUM_HASH_FUNCTIONS
            

            for user in stream:
                hvals = myhashs(user)
                for j, h in enumerate(hvals):
                    tz = count_trailing_zeros(h)
                    if tz > R[j]:
                        R[j] = tz
            

            estimates = [2 ** r for r in R]

            group_size = 10
            group_avgs = []
            for k in range(0, len(estimates), group_size):
                group = estimates[k:k+group_size]
                group_avg = sum(group) / float(len(group))
                group_avgs.append(group_avg)
            group_avgs.sort()
            
            # take the median of the group averages:
            n = len(group_avgs)
            if n % 2 == 1:
                fm_estimate = group_avgs[n // 2]
            else:
                fm_estimate = (group_avgs[n // 2 - 1] + group_avgs[n // 2]) / 2.0
            
            fm_estimate = int(round(fm_estimate))

            total_estimation += fm_estimate
            
            f_out.write(f"{i},{ground_truth},{fm_estimate}\n")

        
        # Calculate and print the overall ratio.
        if total_ground_truth > 0:
            overall_ratio = total_estimation / total_ground_truth
            print(f"\n(sum of estimations / sum of ground truths): {overall_ratio}")
        else:
            print("\nNo ground truth values were accumulated.")

if __name__ == "__main__":
    input_file = sys.argv[1]
    stream_size = int(sys.argv[2])
    num_of_asks = int(sys.argv[3])
    output_file = sys.argv[4]   
    main(input_file, stream_size, num_of_asks, output_file)
